File: agent-api/app/healthcare/tools/rxnorm.py
"""RxNorm Drug Lookup and Interaction Tool"""

import httpx
from typing import Dict, Any, List
from .base import HealthcareTool


class DrugInteractionTool(HealthcareTool):
    name = "check_drug_interaction"
    description = "Check for drug interactions, lookup drug information via RxNorm. Use for medication safety checks."
    cache_ttl = 86400 * 7  # 7 days

    @property
    def input_schema(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "drugs": {
                    "type": "array",
                    "items": {"type": "string"},
                    "description": "List of drug names to check for interactions"
                },
                "drug_name": {
                    "type": "string",
                    "description": "Single drug name to look up"
                }
            }
        }

    async def execute(
        self,
        drugs: List[str] = None,
        drug_name: str = None
    ) -> Dict[str, Any]:
        """Query RxNorm for drug info and interactions"""

        results = {
            "drugs": [],
            "interactions": [],
            "source": "RxNorm / RxNav",
            "source_url": "https://rxnav.nlm.nih.gov/"
        }

        async with httpx.AsyncClient() as client:
            # Get RxCUI for each drug
            drug_list = drugs if drugs else ([drug_name] if drug_name else [])
            rxcuis = []

            for drug in drug_list:
                try:
                    # Search for drug
                    url = "https://rxnav.nlm.nih.gov/REST/drugs.json"
                    response = await client.get(url, params={"name": drug}, timeout=10)
                    data = response.json()

                    concept_group = data.get("drugGroup", {}).get("conceptGroup", [])
                    for group in concept_group:
                        props = group.get("conceptProperties", [])
                        if props:
                            rxcui = props[0].get("rxcui")
                            name = props[0].get("name")
                            if rxcui:
                                rxcuis.append(rxcui)
                                results["drugs"].append({
                                    "name": name,
                                    "rxcui": rxcui,
                                    "original_search": drug
                                })
                            break
                except Exception as e:
                    results["drugs"].append({
                        "name": drug,
                        "error": str(e)
                    })

            # Check interactions if we have multiple drugs
            if len(rxcuis) >= 2:
                try:
                    url = "https://rxnav.nlm.nih.gov/REST/interaction/list.json"
                    response = await client.get(
                        url,
                        params={"rxcuis": "+".join(rxcuis)},
                        timeout=10
                    )
                    data = response.json()

                    interaction_groups = data.get("fullInteractionTypeGroup", [])
                    for group in interaction_groups:
                        for itype in group.get("fullInteractionType", []):
                            for pair in itype.get("interactionPair", []):
                                results["interactions"].append({
                                    "description": pair.get("description"),
                                    "severity": pair.get("severity", "Unknown"),
                                    "drugs": [
                                        c.get("minConceptItem", {}).get("name")
                                        for c in pair.get("interactionConcept", [])
                                    ]
                                })
                except Exception as e:
                    results["interaction_error"] = str(e)

            results["interaction_count"] = len(results["interactions"])
            return results

File: supabase/functions/_shared/healthcare-tools/rxnorm.ts
// RxNorm Drug Interaction Tool

export interface DrugInfo {
  rxcui: string;
  name: string;
  synonym?: string;
  tty?: string;
}

export interface DrugInteraction {
  drug1: string;
  drug2: string;
  severity?: string;
  description: string;
  source: string;
}

export interface RxNormResponse {
  drugs: DrugInfo[];
  interactions: DrugInteraction[];
  count: number;
  source: string;
  source_url: string;
  error?: string;
}

async function getRxCUI(drugName: string): Promise<string | null> {
  try {
    const url = `https://rxnav.nlm.nih.gov/REST/rxcui.json?name=${encodeURIComponent(drugName)}`;
    const response = await fetch(url);
    const data = await response.json();
    return data?.idGroup?.rxnormId?.[0] || null;
  } catch {
    return null;
  }
}

async function getDrugInfo(rxcui: string): Promise<DrugInfo | null> {
  try {
    const url = `https://rxnav.nlm.nih.gov/REST/rxcui/${rxcui}/properties.json`;
    const response = await fetch(url);
    const data = await response.json();
    const props = data?.properties;
    if (props) {
      return {
        rxcui: props.rxcui,
        name: props.name,
        synonym: props.synonym,
        tty: props.tty,
      };
    }
    return null;
  } catch {
    return null;
  }
}

// Alias for compatibility
export const checkDrugInteractions = checkDrugInteraction;

export async function checkDrugInteraction(params: {
  drugs?: string[];
  drug_name?: string;
}): Promise<RxNormResponse> {
  const { drugs, drug_name } = params;

  const drugList = drugs || (drug_name ? [drug_name] : []);

  if (drugList.length === 0) {
    return {
      drugs: [],
      interactions: [],
      count: 0,
      source: 'RxNav',
      source_url: 'https://rxnav.nlm.nih.gov/',
      error: 'At least one drug name is required',
    };
  }

  try {
    // Get RxCUIs for all drugs
    const rxcuis: string[] = [];
    const drugInfos: DrugInfo[] = [];

    for (const drugName of drugList) {
      const rxcui = await getRxCUI(drugName);
      if (rxcui) {
        rxcuis.push(rxcui);
        const info = await getDrugInfo(rxcui);
        if (info) {
          drugInfos.push(info);
        }
      }
    }

    // Check interactions if we have 2+ drugs
    const interactions: DrugInteraction[] = [];

    if (rxcuis.length >= 2) {
      const interactionUrl = `https://rxnav.nlm.nih.gov/REST/interaction/list.json?rxcuis=${rxcuis.join('+')}`;
      const response = await fetch(interactionUrl);
      const data = await response.json();

      const interactionGroups = data?.fullInteractionTypeGroup || [];
      for (const group of interactionGroups) {
        for (const interactionType of group.fullInteractionType || []) {
          for (const pair of interactionType.interactionPair || []) {
            const concepts = pair.interactionConcept || [];
            if (concepts.length >= 2) {
              interactions.push({
                drug1: concepts[0]?.minConceptItem?.name || 'Unknown',
                drug2: concepts[1]?.minConceptItem?.name || 'Unknown',
                severity: pair.severity,
                description: pair.description || 'Interaction detected',
                source: group.sourceName || 'RxNav',
              });
            }
          }
        }
      }
    }

    return {
      drugs: drugInfos,
      interactions,
      count: interactions.length,
      source: 'RxNav',
      source_url: 'https://rxnav.nlm.nih.gov/',
    };
  } catch (error) {
    return {
      drugs: [],
      interactions: [],
      count: 0,
      source: 'RxNav',
      source_url: 'https://rxnav.nlm.nih.gov/',
      error: error instanceof Error ? error.message : 'Unknown error',
    };
  }
}

// OpenFDA Drug Info Lookup
const FDA_API = 'https://api.fda.gov/drug/label.json';

export interface FDADrugInfo {
  success: boolean;
  data?: {
    brand_name?: string;
    generic_name?: string;
    manufacturer?: string;
    route?: string;
    indications?: string;
    contraindications?: string;
    warnings?: string;
    boxed_warning?: string;
  };
  error?: string;
  source: string;
  source_url: string;
}

export async function getDrugInfoFromFDA(params: {
  drug_name: string;
}): Promise<FDADrugInfo> {
  const { drug_name } = params;

  if (!drug_name) {
    return {
      success: false,
      error: 'drug_name is required',
      source: 'OpenFDA',
      source_url: 'https://open.fda.gov/'
    };
  }

  try {
    const query = `openfda.brand_name:"${drug_name}"+OR+openfda.generic_name:"${drug_name}"`;
    const url = `${FDA_API}?search=${encodeURIComponent(query)}&limit=1`;

    const response = await fetch(url);
    const data = await response.json();

    if (!data.results || data.results.length === 0) {
      return {
        success: false,
        error: 'Drug not found in FDA database',
        source: 'OpenFDA',
        source_url: 'https://open.fda.gov/'
      };
    }

    const drug = data.results[0];
    const openfda = drug.openfda || {};

    return {
      success: true,
      data: {
        brand_name: openfda.brand_name?.[0],
        generic_name: openfda.generic_name?.[0],
        manufacturer: openfda.manufacturer_name?.[0],
        route: openfda.route?.[0],
        indications: drug.indications_and_usage?.[0]?.slice(0, 500),
        contraindications: drug.contraindications?.[0]?.slice(0, 500),
        warnings: drug.warnings?.[0]?.slice(0, 500),
        boxed_warning: drug.boxed_warning?.[0]?.slice(0, 500)
      },
      source: 'OpenFDA Drug Labels',
      source_url: 'https://open.fda.gov/'
    };

  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error',
      source: 'OpenFDA',
      source_url: 'https://open.fda.gov/'
    };
  }
}
